config.org
#+TITLE: config.el
#+PROPERTY: header-args :tangle yes

#+BEGIN_SRC elisp
  (with-eval-after-load 'org
#+END_SRC

* Table of Contents :TOC_4_gh:
- [[#encryption][Encryption]]
  - [[#for-example][For example]]
- [[#look-and-feel][Look and feel]]
- [[#journal][Journal]]
- [[#keywords][Keywords]]
- [[#link-abbreviations][Link abbreviations]]
- [[#capture][Capture]]
  - [[#overview][Overview]]
  - [[#templates][Templates]]
    - [[#generic-meeting][Generic Meeting]]
    - [[#cfengine-customer-status-check-in][CFEngine Customer Status Check-in]]
    - [[#cfengine-sales][CFEngine Sales]]
    - [[#cfengine-backlog-grooming][CFEngine Backlog Grooming]]
    - [[#cfengine-planning][CFEngine Planning]]
    - [[#cfengine-standup][CFEngine Standup]]
    - [[#the-agent-is-in][The Agent is in]]
    - [[#security-team][Security Team]]
    - [[#sre-team][SRE Team]]
    - [[#unorganized][Unorganized]]
      - [[#fix-myorg-daily-review-capture-template-yesterday-so-that-it-inserts-yesterdays-date-in-the-headline][Fix =my/org-daily-review-capture-template-yesterday= so that it inserts yesterdays date in the headline]]
  - [[#configure-capture][Configure Capture]]
    - [[#daily-review][Daily Review]]
    - [[#meetings][Meetings]]
    - [[#capturing-under-currently-clocked-task][Capturing under currently clocked task]]
    - [[#capture-to-point][Capture to point]]
    - [[#capture-new-entry-to-org-journal][Capture new entry to org-journal]]
    - [[#pipe-into-example-block-from-outside-org-mode][Pipe into EXAMPLE block from outside org-mode]]
    - [[#selection-outside-emacs-into-currently-clocked-task][Selection (outside emacs) into currently clocked task]]
- [[#babel-src-blocks][Babel (SRC blocks)]]
- [[#exporting][Exporting]]
  - [[#publishing][Publishing]]
- [[#google-calendar-sync][Google Calendar Sync]]
  - [[#org-cal][org-cal]]
- [[#agenda][Agenda]]
- [[#clocking][Clocking]]
- [[#downloading][Downloading]]
  - [[#pictures][Pictures]]
- [[#jira][Jira]]
- [[#for-the-future][For the future]]
- [[#eof][EOF]]

* Encryption
:PROPERTIES:
:ID:       b4bded6c-9d3a-4570-9cc0-ae011226399b
:END:
 
GPG can be used to automatically encrypt sections of text tagged with /*crypt*/.

#+BEGIN_SRC elisp
  (require 'org-crypt)
  (org-crypt-use-before-save-magic)
  (setq org-tags-exclude-from-inheritance (quote ("crypt")))
  (setq org-crypt-key "9274E588E866A10B713C9CCD9EB3AD425D1CCC11")
#+END_SRC

Sometimes, when I am at home, I will ssh into my laptop and run emacs at the shell, or attach to a running server with emacsclient.
I have had issues with gpg poping up a gui and then I can't decrupt from my ssh session. This resolves that for me.

#+BEGIN_SRC elisp
  (setq epa-pinentry-mode 'loopback)
  ;; not working (pinentry-start)
#+END_SRC

#+RESULTS:
: loopback

** For example  :crypt:
-----BEGIN PGP MESSAGE-----

hQIMA/jYNcPWXvJ6AQ/+ItpAIG1h5g31OxHivWHoEBnnKpOILl9bvbtv+rjxtRFL
aJxKFK6SL7iKdFogKyzU5EpS9qF9aRSH9eEivaiCcF8uYIbDQkU1EW5vadcUbpx5
oKm2JNy4QYy/Yj5709+gxO+zIITXN4Mo5CuXF9OKLp0euuwLm2bZkYrvjhryGLqe
lS3+bovvzX9VKdYYE5Bdulq5aTuVspOwpUhWcEJSql+Ittry9stpbeUa654ikEEn
UR5RVreZujYLQ4s0/+A1/kGAdp+3U+roef9FCcAH75eFkvw7OsLcw7YPyviCdPs0
lroZAYDfxYV0332Ak2nj+cMxdwm/lRPXQfqWJ+XJUxQadwaLjbsrqShZBOjjf3eP
6fXQHauLn0twJ1jAR5RePE+AqxHiY3b4Revfs+FwnBal5N+fiV5QaVSnEpbhGoWk
UrMKbm1yz0PP2vnm72j0nhZzf5uaT2MRItL2Fb2hDJexUaO745WqP69uk9kpdxm7
RHM/KDO1/g1gFt6XXDlFo+XzeaByhP6mEhbna3tk13660UR03CKRieYew4K3bJHJ
XUerJJ9O+mRwq1x/Fd22FUEs8ZOzuRHLWlRjfl9aNN7AK5UGCxANV3SEkc/I+HgM
vhdT3bWoYIjMQQCAFXMkLhXxzZKCr9W6iOWhIskKuFbyckxlmuC2KmeiP2tpO7/S
SgE1eqBxX6fpIM0Pzp2e5/XqB9tW6QxKavgBsCGu521UqS+XQTfX9K/Z0cFnpVsW
tJOSK/O5zYB2p4vGLcelSWN1+kkMgsgDwPv9
=AGgn
-----END PGP MESSAGE-----

* Look and feel
:PROPERTIES:
:ID:       7d815734-bf0e-4620-89e6-910416ec7cc6
:END:

#+BEGIN_SRC elisp
  ;; BEGIN org niceities
  ;; NOT WORKING!
  (setq-default dotspacemacs-large-file-size 3)
  (setq org-tags-match-list-sublevels 'indented)
  (setq org-startup-indented t)
  (setq org-startup-with-inline-images t)

  ;; Familiar zooming with Ctrl+ and Ctrl-
  (define-key global-map (kbd "C-+") 'text-scale-increase)
  (define-key global-map (kbd "C--") 'text-scale-decrease)

  (setq org-insert-heading-respect-content t)
  (setq org-id-link-to-org-use-id t)
  (setq org-id-track-globally t)

  ;; TODO Move this out from the org-layer, it's not org-specific.
  ;; Or make it specific to my org-files somehow https://emacs.stackexchange.com/a/3611
  (setq global-auto-revert-mode t)

  ;; Disable line highlighting in terminal mode (makes it hard to see)
  ;;(if display-graphic-p (global-hl-line-mode nil))
  ;; END org niceities
#+END_SRC

Variable pitch, and visual line mode for a better writing environment.
https://zzamboni.org/post/beautifying-org-mode-in-emacs/
#+CAPTION: variable pitch and visual line mode
#+BEGIN_SRC elisp
  (custom-theme-set-faces
   'user
   '(org-block                 ((t (:inherit fixed-pitch))))
   '(org-document-info         ((t (:foreground "dark orange"))))
   '(org-document-info-keyword ((t (:inherit (shadow fixed-pitch)))))
   '(org-link                  ((t (:foreground "royal blue" :underline t))))
   '(org-meta-line             ((t (:inherit (font-lock-comment-face fixed-pitch)))))
   '(org-property-value        ((t (:inherit fixed-pitch))) t)
   '(org-table                 ((t (:inherit fixed-pitch))) t) ;; Much improved, but right side of table still doesn't align
   '(org-verbatim              ((t (:inherit (shadow fixed-pitch)))))
   '(org-code                  ((t (:inherit (shadow fixed-pitch)))))
   '(org-special-keyword       ((t (:inherit (font-lock-comment-face fixed-pitch)))))
   '(org-tag                   ((t (:inherit (shadow fixed-pitch) :weight bold :height 0.8))))
   '(org-verbatim              ((t (:inherit (shadow fixed-pitch)))))
   '(org-indent                ((t (:inherit (org-hide fixed-pitch))))))
  (add-hook 'org-mode-hook 'variable-pitch-mode)
  (add-hook 'org-mode-hook 'visual-line-mode)
#+END_SRC

* Journal
:PROPERTIES:
:ID:       778fcfc3-5ca1-4af1-86a1-654c97260f3a
:END:

These need to be set earlier, like with the layer specification, for example:

#+BEGIN_SRC elisp :tangle no
  (org :variables
    org-enable-reveal-js-support t
    org-enable-org-journal-support t)
#+END_SRC 

#+BEGIN_SRC elisp
  (setq org-journal-dir "~/org/journal/")
  (setq org-journal-file-format "%Y-%m-%d")
  (setq org-journal-date-prefix "#+TITLE: ")
  (setq org-journal-date-format "%A, %B %d %Y")
  (setq org-journal-time-prefix "* ")
  (setq org-journal-time-format "")
  (setq org-want-todo-bindings t)
#+END_SRC
  
* ToDO Keywords
:PROPERTIES:
:ID:       9f7dc47a-843d-4d2d-8178-3053c57c30bf
:END:
  
#+BEGIN_SRC elisp
  ;; TODO Keywords
  (setq org-todo-keywords
        (quote ((sequence "TODO(t)" "IN_PROGRESS(i)" "|" "DONE(d)")
                (sequence "WAITING(w@/)" "DELEGATED(D)" "HOLD(h@/)" "|" "CANCELLED(c@/)"))))

  ;; This conflicts with the evil key bindings in spacemacs, instead of using the old speedcommands, use =, T T= when inside org mode.
  ;;(setq org-use-fast-todo-selection t)
  ;;(setq org-use-speed-commands t)

#+END_SRC

* Link abbreviations
:PROPERTIES:
:ID:       2d641838-b9f5-4f89-a228-0bd570dab880
:END:
:(PROPERTIES):
:ID:       83e904da-0e44-4629-942f-c3a390e227fb
:END:
  
#+BEGIN_SRC elisp
      ;; Link abbreviations http://orgmode.org/manual/Link-abbreviations.html#Link-abbreviations
      ;; This makes it easy to create links in org files to common urls
      ;; Note: The actual link is not stored in the text, only when rendered
      ;; Usage: [[zendesk:2753]] or [[redmine:7481][My text]]
      (setq org-link-abbrev-alist
            '(
              ("cfe-function" . "https://docs.cfengine.com/docs/master/reference-functions-%s.htm")
              ("cfe-promise" . "https://docs.cfengine.com/docs/3.15/reference-promise-types-%s.htm")
              ("zendesk" . "https://cfengine.zendesk.com/agent/tickets/")
              ("redmine" . "https://dev.cfengine.com/issues/")
              ("core-pr" . "https://github.com/cfengine/core/pull/")
              ("mpf-pr" . "https://github.com/cfengine/masterfiles/pull/")
              ("core-commit" . "https://github.com/cfengine/core/commit/")
              ("mpf-commit" . "https://github.com/cfengine/masterfiles/commit/")
              ("jira" . "https://tracker.mender.io/browse/")))
#+END_SRC

* Capture :ATTACH:
:PROPERTIES:
:ID:       c6f828da-99c2-4354-80ef-9ab57c9f702d
:Attachments: 2018-11-13-26-27_2018-11-02_13-27-56.png 2018-11-13-26-50_2018-11-02_13-28-14.png
:END:
** Overview
I use capture templates to help me get information for common things in a
consistent way. I have capture templates for various types of meetings, and
different types of tasks that I work on.

#+DOWNLOADED: file:///home/nickanderson/Pictures/Screenshots/2018-11-13-26-27.png @ 2018-11-02 13:27:56
[[file:data/c6/f828da-99c2-4354-80ef-9ab57c9f702d/2018-11-13-26-27_2018-11-02_13-27-56.png]]


#+DOWNLOADED: file:///home/nickanderson/Pictures/Screenshots/2018-11-13-26-50.png @ 2018-11-02 13:28:14
[[file:data/c6/f828da-99c2-4354-80ef-9ab57c9f702d/2018-11-13-26-50_2018-11-02_13-28-14.png]]
** Templates
Templates can be external, probably it would be a good idea to try moving to
external template files instead of having huge templates in elisp. [[https://www.reddit.com/r/emacs/comments/7zqc7b/share_your_org_capture_templates/duqzj0r/][This post
from reddit.com/emacs]] shows an example using external templates.

*** Generic Meeting
:PROPERTIES:
:ID:       4b0d8c14-b69a-4956-8874-10b3d7641c99
:END:
#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-generic "* %u %^{Meeting about} %^g
  CREATED: %U

  %?

  ,** Notes


  " "Meeting Template - Generic")
#+END_SRC

*** CFEngine Customer Status Check-in
:PROPERTIES:
:ID:       6dcf9681-425f-4bdc-8ded-c72f3ada392d
:END:
#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-customer-status "* %u Status Check-in
  :PROPERTIES:
  %^{CUSTOMER}p
  :CREATED: %U
  :MAIL_TO: Team CFEngine <cfengine@northern.tech>
  :MAIL_SUBJECT: Meeting Notes: CFEngine Customer Sync
  :MAIL_FROM: Nick Anderson <nick.anderson@northern.tech>
  :END:

  %?

  ,*Attendees:*
    - Nick Anderson
    -

  ,** Info
  - Current versions running:
  - Next planned upgrade:
  - Upcoming platform changes:

  ,** Notes

  " "Meeting Template - Customer Status Check-in")
#+END_SRC

*** CFEngine Sales
:PROPERTIES:
:ID:       eefc7827-aa08-4c11-b68c-9ff0fa8a6331
:END:
#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-cfengine-sales "* %u CFEngine Sales 
  %^{CUSTOMER}pCREATED: %U

  %?

  ,*Attendees:*
    - Nick Anderson
    -

  ,** Info
  - Platforms:
  - Infrastructure Size:
  - Key use cases:

  ,** Notes

  " "Meeting Template - CFEngine Sales")
#+END_SRC

*** CFEngine Backlog Grooming
:PROPERTIES:
:ID:       5f77724f-afa8-47bc-a858-32396a37472c
:END:

Bi-weekly meeting to make sure we are ready to work on our planned tasks.

*Agenda:*

- [[https://tracker.mender.io/issues/?filter=11300][Review New customer issues]]
- [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=34][Review Progress on CFEngine Epics]]
- [[https://tracker.mender.io/issues/?filter=11205][Review Understanding of Next Bugs]]
- [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=11&view=planning&epics=visible][Review CFEngine PM Backlog]]

#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-grooming "* %u Meeting About CFEngine Grooming :internal_meeting:
  :PROPERTIES:
  :CREATED: %U
  :END:
  %?
  ,*Agenda:*
  - [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=49&view=planning&selectedIssue=ENT-5265&epics=visible&issueLimit=100][CFEngine Epics]] (click the green/gray bars for summary)
  - [[https://tracker.mender.io/issues/?filter=11300][Review New customer issues]]
  - [[https://tracker.mender.io/issues/?filter=11205][Review Understanding of Next auto-prioritized Bugs]]
  - [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=26&view=planning.nodetail&epics=visible&issueLimit=100][Review Backlog]]

  ,** Notes

  " "Meeting Template - Grooming")
#+END_SRC

*** CFEngine Planning
:PROPERTIES:
:ID:       113a8e92-53f7-4688-be4d-aa17cbed1512
:END:

#+BEGIN_SRC elisp
      ;; BEGIN Capture Templates
      ;; I picked up this neat trick from the Venerable Sacha Chua
      (defvar my/org-meeting-template-planning "* %u Meeting About CFEngine Planning
  CREATED: %U

  - [[https://docs.google.com/document/d/1OY8RgkRuUghr9rImrx7kzeHJAPlZw9XY5Is1sGf5Rt0/edit?usp=drive_web&ouid=105225911714445244426][CFEngine SCRUM log]]

  ,*Agenda:*
   - Review
     - Demos
   - Review work in progress
   - Review newly registered issues

  ,** Review
  ,*** Demos
  ,*** Leftovers

  ,#+CAPTION: Summary of story points remaining at end of sprint
  |---------+------------------------|
  |          | Story Points Remaining |
  |----------+------------------------|
  | UI       |                      0 |
  | Core     |                      0 |
  | Enablers |                      0 |
  |----------+------------------------|
  | *Total*  |                      0 |
  ,#+TBLFM: @3$2=remote(core-leftovers,$total)::@2$2=remote(ui-leftovers,$total)::@4$2=remote(enablers-leftovers,$total)::@5$2=vsum(@2..@-1)

  ,**** UI

  ,#+CAPTION: UI related story points left at end of sprint
  ,#+name: ui-leftovers
  |----------+--------------+-------------------------------------------------------|
  | Ticket   | SP Remaining | Comment                                               |
  |----------+--------------+-------------------------------------------------------|
  |          |              |                                                       |
  |----------+--------------+-------------------------------------------------------|
  | *Total*  |            0 |                                                       |
  | ^        |        total |                                                       |
  ,#+TBLFM: $total=vsum(@2..@-1)

  ,**** Core

  ,#+CAPTION: Core related story points left at end of sprint
  ,#+name: core-leftovers
  |----------+--------------+-------------------------------------------------------|
  | Ticket   | SP Remaining | Comment                                               |
  |----------+--------------+-------------------------------------------------------|
  |          |              |                                                       |
  |----------+--------------+-------------------------------------------------------|
  | *Total*  |            0 |                                                       |
  | ^        |        total |                                                       |
  ,#+TBLFM: $total=vsum(@2..@-1)

  ,**** Enablers

  ,#+CAPTION: Enablers related story points left at end of sprint
  ,#+name: enablers-leftovers
  |----------+--------------+-------------------------------------------------------|
  | Ticket   | SP Remaining | Comment                                               |
  |----------+--------------+-------------------------------------------------------|
  |          |              |                                                       |
  |----------+--------------+-------------------------------------------------------|
  | *Total*  |            0 |                                                       |
  | ^        |        total |                                                       |
  ,#+TBLFM: $total=vsum(@2..@-1)

  ,** Planning
  - [[https://tracker.mender.io/secure/Dashboard.jspa?selectPageId=11201][Sprint status]]
  - [[https://docs.google.com/spreadsheets/d/12WHg2SlP6xZ5oG0Dnw2KsOrAkDjeJ4o92osLBr_kbPc/edit?usp=drive_web&ouid=105225911714445244426][CFEngine Capacity]]
  - [[https://tracker.mender.io/issues/?filter=11205][Magical priority bug list]]
  - [[https://tracker.mender.io/issues/?filter=11300][New Customer issues in last two weeks]]
  - [[https://tracker.mender.io/issues/?filter=11211][Recently active customer related issues]]
  - [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=49&view=detail][CFEngine SCRUM board]]
  - [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=26][CFEngine Kanban board]]


  - New issues affecting customers
  - New issues not necessarily affecting customers

  " "Meeting Template - Planning")
#+END_SRC

*** CFEngine Standup
:PROPERTIES:
:ID:       fdeefc35-4d8d-4b94-95d1-2126bbabfe3b
:END:

#+CAPTION: Capture template for CFEngine Standup Meeting
#+BEGIN_SRC elisp
  (defun my/northern-tech-org-files ()
  "Return list of .org files containing information related to northern.tech. ( ~/org/refile.org, ~org/Work/northern.tech/, excluding temporary files (.# prefix)"

  (cl-remove-if (lambda (k)
                  ;; We remove temporary files (files containing .#) because it results in an
                  ;; error trying trying to generate a clocktable from the resulting list as that
                  ;; file appears and disappears.
                  (string-match "\\.#" k))
                (append
                 ;; Captures and most recent history is in my refile.org
                 '("~/org/refile.org")
                 ;; And my northern.tech specific tree
                 (directory-files-recursively "~/org/Work/northern.tech" "[^.]\\.*org$" nil))))

  (defvar my/org-meeting-template-standup "* %u CFEngine Standup  :internal_meeting:
  :PROPERTIES:
  :CREATED: %U
  :MAIL_TO:  cfengine@northern.tech
  :MAIL_SUBJECT: Standup Meeting: CFEngine
  :MAIL_FROM: Nick Anderson <nick.anderson@northern.tech>
  :END:

  ,#+BEGIN_NOTE
    Did you know, you can find a [[https://gitlab.com/nickanderson/northern.tech/blob/master/standup-meetings.org][history of standup meeting notes here]]?
  ,#+END_NOTE
  %?
  ,** Aleksei
  ,** Craig
  ,** Igor
  ,** Nick

  ,#+begin_src elisp :results list
    (org-ql-query
      :select '(org-get-heading t t)
      :where '(ts :from -1 :to today)
      :from (my/northern-tech-org-files)
       )
  ,#+end_src

  ,#+BEGIN: clocktable :stepskip0 t :fileskip0 t :maxlevel 2 :scope my/northern-tech-org-files :block yesterday :tags t :link nil :narrow 200!\n#+END

  ,** Ole
  ,** Vratislav

    " "Meeting Template - CFEngine Standup")
#+END_SRC

*** The Agent is in
:PROPERTIES:
:ID:       e6a1c237-8588-4c05-9210-19dbedc52f2e
:END:
#+CAPTION: Capture template for The Agent is in Meeting
#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-the-agent-is-in "* %u The Agent is in :meeting:
  :PROPERTIES:
  :CREATED: %U
  :MAIL_TO:  people@northern.tech
  :MAIL_SUBJECT: %u The Agent is in
  :MAIL_FROM: Nick Anderson <nick.anderson@northern.tech>
  :END:

  ,#+BEGIN_NOTE
    Did you know, you can find a [[https://gitlab.com/nickanderson/northern.tech/blob/master/CFEngine/the-agent-is-in.org][history of The Agent is in  meetings]]?
  ,#+END_NOTE

  ,** Topics

    " "Meeting Template - The Agent is in")
#+END_SRC

*** Security Team
:PROPERTIES:
:ID:       e08e3a50-2fd7-4670-944e-7e72b90de350
:END:
#+CAPTION: Capture template for Security Team Meeting
#+BEGIN_SRC elisp
  (defvar my/org-meeting-template-security-team "* %u Security Team meeting notes
  :PROPERTIES:
  :CREATED: %U
  :MAIL_TO:  people@northern.tech
  :MAIL_SUBJECT: %u Security Team meating notes
  :MAIL_FROM: Nick Anderson <nick.anderson@northern.tech>
  :END:

  ,** Resources
  - [[https://docs.google.com/document/d/1KRwilxJlb8Qj4bkogn9ScTJf1MWh5thaAEEEFN-GoTw][Meeting notes and agenda]]
  - [[https://docs.google.com/spreadsheets/d/19ka3VJ2BwjjL_iYBNUre1lss_p7oBxZpXV2jklr1etc/][Working document]]
  - [[https://northern-tech.slack.com/archives/CT9ALP468][Slack channel]]

config.el
(with-eval-after-load 'org

(require 'org-crypt)
(org-crypt-use-before-save-magic)
(setq org-tags-exclude-from-inheritance (quote ("crypt")))
(setq org-crypt-key "9274E588E866A10B713C9CCD9EB3AD425D1CCC11")

;; BEGIN org niceities
;; NOT WORKING!
(setq-default dotspacemacs-large-file-size 3)
(setq org-tags-match-list-sublevels 'indented)
(setq org-startup-indented t)
(setq org-startup-with-inline-images t)

;; Familiar zooming with Ctrl+ and Ctrl-
(define-key global-map (kbd "C-+") 'text-scale-increase)
(define-key global-map (kbd "C--") 'text-scale-decrease)

(setq org-insert-heading-respect-content t)
(setq org-id-link-to-org-use-id t)
(setq org-id-track-globally t)

;; TODO Move this out from the org-layer, it's not org-specific.
;; Or make it specific to my org-files somehow https://emacs.stackexchange.com/a/3611
(setq global-auto-revert-mode t)
;; END org niceities

(setq org-journal-dir "~/org/journal/")
(setq org-journal-file-format "%Y-%m-%d")
(setq org-journal-date-prefix "#+TITLE: ")
(setq org-journal-date-format "%A, %B %d %Y")
(setq org-journal-time-prefix "* ")
(setq org-journal-time-format "")
(setq org-want-todo-bindings t)

;; TODO Keywords
(setq org-todo-keywords
      (quote ((sequence "TODO(t)" "IN_PROGRESS(i)" "|" "DONE(d)")
              (sequence "WAITING(w@/)" "DELEGATED(D)" "HOLD(h@/)" "|" "CANCELLED(c@/)"))))

;; This conflicts with the evil key bindings in spacemacs, instead of using the old speedcommands, use =, T T= when inside org mode.
;;(setq org-use-fast-todo-selection t)
;;(setq org-use-speed-commands t)

;; Link abbreviations http://orgmode.org/manual/Link-abbreviations.html#Link-abbreviations
;; This makes it easy to create links in org files to common urls
;; Note: The actual link is not stored in the text, only when rendered
;; Usage: [[zendesk:2753]] or [[redmine:7481][My text]]
(setq org-link-abbrev-alist
      '(
        ("cfe-func" . "https://docs.cfengine.com/docs/master/reference-functions-")
        ("zendesk" . "https://cfengine.zendesk.com/agent/tickets/")
        ("redmine" . "https://dev.cfengine.com/issues/")
        ("core-pr" . "https://github.com/cfengine/core/pull/")
        ("mpf-pr" . "https://github.com/cfengine/masterfiles/pull/")
        ("core-commit" . "https://github.com/cfengine/core/commit/")
        ("mpf-commit" . "https://github.com/cfengine/masterfiles/commit/")
        ("jira" . "https://tracker.mender.io/browse/")))

(defvar my/org-meeting-template-generic "* %u %^{Meeting about} %^g
CREATED: %U

%?

** Notes


" "Meeting Template - Generic")

(defvar my/org-meeting-template-customer-status "* %u Status Check-in
%^{CUSTOMER}pCREATED: %U

%?

*Attendees:*
  - Nick Anderson
  -

** Info
- Current versions running:
- Next planned upgrade:
- Upcoming platform changes:

** Notes

" "Meeting Template - Customer Status Check-in")

(defvar my/org-meeting-template-grooming "* %u Meeting About CFEngine Grooming :internal_meeting:
CREATED: %U
%?
*Agenda:*
- [[https://tracker.mender.io/issues/?filter=11300][Review New customer issues]]
- [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=34][Review Progress on CFEngine Epics]]
- [[https://tracker.mender.io/issues/?filter=11205][Review Understanding of Next Bugs]]
- [[https://tracker.mender.io/secure/RapidBoard.jspa?rapidView=11&view=planning&epics=visible][Review CFEngine PM Backlog]]

** Notes

" "Meeting Template - Grooming")

;; BEGIN Capture Templates
    ;; I picked up this neat trick from the Venerable Sacha Chua
    (defvar my/org-meeting-template-planning "* %u Meeting About CFEngine Planning   %^G
CREATED: %U

*Agenda:*
 - Demos
 - Review work in progress
 - Review newly registered issues

** Demos


" "Meeting Template - Planning")

    (defvar my/org-meeting-template-standup "* %u CFEngine Standup  :internal_meeting:
CREATED: %U

** Aleksei
** Igor
** Ole
** Vratislav
** Nils 
** Craig
** Nick

" "Meeting Template - Standup")

    (defvar my/org-meeting-template "* %u %^{Meeting About ...}   %^G
CREATED: %U

*Attendees:*

 - [X] Nick Anderson
 - [ ] %?


*Agenda:*
 -
 -

*Notes:*


" "Meeting Template")

    (defvar my/org-contact-capture-template "* %(org-contacts-template-name)
CREATED: %U
:PROPERTIES:
:EMAIL: %(org-contacts-template-email)
:END:")


    (defvar my/org-respond-email-capture-template "** TODO [#B] Respond to %:from on %:subject\nSCHEDULED: %t\n%U\n%a\n\n")

    (defvar my/org-capture-support "** TODO [#A] [[zendesk:%^{ISSUE}]]: %^{DESCRIPTION} %^G\n\n%?\n")
    (defvar my/org-capture-jira "** TODO [#B] [[jira:%^{ISSUE}]]: %^{DESCRIPTION} %^G\n\n%?\n")


    (defvar my/org-daily-review-capture-template "* %u\n\n%?\n")

;; Configure custom capture templates
(setq org-capture-templates
      `(;; Note the backtick here, it's required so that the defvar based tempaltes will work!
        ;;http://comments.gmane.org/gmane.emacs.orgmode/106890

        ("t" "To-do" entry (file+headline "~/org/refile.org" "Tasks")
         "** TODO %^{Task Description}\nCreated From: %a\n%?\n\n" :clock-in t :clock-resume t :append t)

        ("s" "Support" entry (file+headline "~/org/refile.org" "Tasks")
         ,my/org-capture-support :clock-in t :clock-resume t :append t)

        ("j" "Jira" entry (file+headline "~/org/refile.org" "Tasks")
         ,my/org-capture-jira :clock-in t :clock-resume t :append t)

        ("w" "Web site" entry
         (file "~/org/websites.org")
         "* %a :website:\n\n%U %?\n\n%:initial" :append t)

        ("r" "Respond to Email" entry (file+headline "~/org/refile.org" "Tasks")
         ,my/org-respond-email-capture-template :clock-in t :clock-resume t :append t)

        ("c" "Contact" entry (file "~/org/x-files.org") ,my/org-contact-capture-template :append t)
        ("d" "Daily Review" entry (file "~/org/daily_reviews.org") ,my/org-daily-review-capture-template :clock-in t :clock-resume t :append t)

        ("m" "Meetings" )
        ("ms" "Meeting - Standup" entry (file "~/org/cfengine/meetings.org" )
         ,my/org-meeting-template-standup :clock-in t :clock-resume t :append t :empty-lines-after 1)
        ("mc" "Meeting - Customer Status Check-in" entry (file "~/org/cfengine/meetings.org" )
         ,my/org-meeting-template-customer-status :clock-in t :clock-resume t :append t :empty-lines-after 1)
        ("mg" "Meeting - Grooming" entry (file "~/org/cfengine/meetings.org" )
         ,my/org-meeting-template-grooming :clock-in t :clock-resume t :append t :empty-lines-after 1)
        ("mp" "Meeting - Planning/Review" entry (file "~/org/cfengine/meetings.org" )
         ,my/org-meeting-template-planning :clock-in t :clock-resume t :append t :empty-lines-after 1)
        ("mm" "Meeting - Generic" entry (file "~/org/cfengine/meetings.org" )
         ,my/org-meeting-template-generic :clock-in t :clock-resume t :append t :empty-lines-after 1)
        ))
;; END Capture templates
;; Use UUIDs to identify each speicifc entry
(add-hook 'org-capture-prepare-finalize-hook 'org-id-get-create)

(setq org-src-fontify-natively t)

(setq org-src-tab-acts-natively t)

(org-babel-do-load-languages
 'org-babel-load-languages
 '(
   (shell . t) ; Works for sh, shell, bash
   ;; (python . t)
   ;; (ruby . t)
   ;; (ditaa . t)
   ;; (http . t)
   ;; (plantuml . t)
   ;; (perl . t))
   ))

;; This is no longer needed. It's handled by the cfengine layer automatically if
;; it sees the org layer is also used.
;; https://github.com/syl20bnr/spacemacs/pull/11528
;; (when (configuration-layer/layer-usedp 'cfengine)
;;   ;;(require 'ob-cfengine3) ;; I have problems with capture templates if I don't
;;   ;; use this before capturing or require it.

;;   (append org-babel-load-languages
;;           '((cfengine3 . t)))
;;   )

;; BEGIN exports
;; Disable exporting subscripts (I use a lot of underscores, and they are never for subscript)
(setq org-export-with-sub-superscripts nil)

(when (configuration-layer/package-usedp 'ox-jira)
  (add-to-list 'org-export-backends 'jira))

(when (configuration-layer/layer-usedp 'markdown)
  (add-to-list 'org-export-backends 'md))

;; END exports

;; BEGIN Publishing
(setq org-publish-project-alist
      '(
        ("journal"
         :base-directory "~/org/journal/"
         :base-extension ""
         :publishing-directory "~/journal/"
         :recursive t
         :publishing-function org-html-publish-to-pdf
         :headline-levels 1
         :autopreamble nil)
        ("cfengine-html"
         :base-directory "~/org/cfengine/"
         :base-extension "org"
         :publishing-directory "~/CFEngine/Google Drive/nicks_org"
         :recursive t
         :publishing-function org-html-publish-to-html
         :headline-levels 4
         :autopreamble t
         :eval "never-export")
        ("cfengine-org"
         :base-directory "~/org/cfengine/"
         :base-extension "org"
         :publishing-directory "~/CFEngine/Google Drive/nicks_org"
         :recursive t
         :publishing-function org-org-publish-to-org
         :headline-levels 4
         :autopreamble t)
        ("cfengine-pdf"
         :base-directory "~/org/cfengine/"
         :base-extension "org"
         :publishing-directory "~/CFEngine/Google Drive/nicks_org"
         :recursive nil
         :publishing-function org-latex-publish-to-pdf
         :headline-levels 4
         :autopreamble t)
        ("cfengine-txt"
         :base-directory "~/org/cfengine/"
         :base-extension "org"
         :publishing-directory "~/CFEngine/Google Drive/nicks_org"
         :recursive t
         :publishing-function org-ascii-publish-to-utf8
         :headline-levels 4
         :autopreamble t)

        ))
;; END Publishing
;; BEGIN org-agenda configuration

(setq org-agenda-span 'day)

(setq org-agenda-files
      '("~/org" "~/org/cfengine" "~/org/cfengine/customers" "~/.org-jira"))
;; It's hard to see them (at least with the default color). Also this is a
;; reccomended change to speed up the agenda (not that it's too slow for me).
(setq org-agenda-dim-blocked-tasks nil)
(setq org-agenda-prefix-format '"%b")
;; END org-agenda configuration

(when (configuration-layer/package-used-p 'org-timeline)
  (require 'org-timeline)
  (add-hook 'org-agenda-finalize-hook 'org-timeline-insert-timeline :append)
  )

;; BEGIN clocking configuration
(setq spaceline-org-clock-p t)
(setq org-clock-idle-time 15)
(setq org-time-clocksum-format (quote (:hours "%d" :require-hours t :minutes ":%02d" :require-minutes t)))
;; END clocking configuration

(setq org-download-method 'attach)

(when (configuration-layer/package-used-p 'org-jira)
           (setq jiralib-url "https://tracker.mender.io:443")
            (setq org-jira-working-dir "~/.org-jira"))

)
